bash-preexec.sh

```diff
@@ -95,9 +99,9 @@ __bp_interactive_mode() {
 # This function is installed as part of the PROMPT_COMMAND.
 # It will invoke any functions defined in the precmd_functions array.
 __bp_precmd_invoke_cmd() {
-    # Save the returned value from our last command. Note: this MUST be the
-    # first thing done in this function.
-    __bp_last_ret_value="$?"
+    # Save the returned value from our last command, and from each process in
+    # its pipeline. Note: this MUST be the first thing done in this function.
+    __bp_last_ret_value="$?" BP_PIPESTATUS=("${PIPESTATUS[@]}")
```

Comment:
By the looks of it `BP_PIPESTATUS` will not be usable from `preexec` because `PIPESTATUS` isn't updated at the time the `DEBUG` trap fires, so it should probably be unset at the end of `__bp_precmd_invoke_cmd` so a stale value isn't propagated.

Reply:
I'm not too familiar with the `DEBUG` trap, but wouldn't the first one after the user hits enter expect to see the same values of `$?` and `$[BP_]PIPESTATUS` as immediately before the user hit enter?